### DISCUSSION TITLE
Logging Console Output

### DISCUSSION THREAD

**Discussion by swatkinson** [cwapi3d/cwapi3dpython#198] [General] [answered]
Hey,

I'm working with some custom plugins and they are causing cadwork to crash whilst I'm working. When I have the Test Options > Console open, the error information I need is printed to the console. Is there a way to log this standard output in the console to a text file?
I see that there is a Create Log-File in Test Options, I'm not sure if this is what I am looking for, but I cannot find the file it is making anyways, and it un-checks itself every time I re-open cadwork. 
I've also been using Start Debug in Test Options, but as far as I've seen in my testing, nothing has been written to it and it remains a blank text file, even on a crash.

If anyone has any insight as to how I can log these errors, I'd greatly appreciate it!
(also, apologies if anything is unclear, I am quite new to cadwork and making cadwork plugins)

**Comment by Brunner246** [comment, marked as answer]
@swatkinson 
Please do not use Start Debug in Test Options. 

You could try something like this 
```python
import sys
import traceback

import utility_controller as uc

PLUGIN_PATH = uc.get_plugin_path()

log_file = open(fr"{PLUGIN_PATH}/log.txt", "w")

sys.stderr = log_file

def raise_error():
    raise ValueError("some test error")

try:
    raise_error()

except Exception as e:
    traceback.print_exc()

log_file.close()
```

Does it crash because of your python code ? Or what causes the crash ?

**Reply by swatkinson** [reply]
Thanks for the suggestion, I'll look into replicating that in my plugins

I'm trying to log the crashes to find out what is causing the issues. Unfortunately I am not the only one using my plugin, so unless everyone has the console open and sends me a screenshot of their issues, I cannot properly investigate the cause of the crashes.
That's why I'm looking into seeing if I can log that console to a log file, so that I can look at the other peoples logs after they experience a crash.

The hunch right now is that if someone tries to do something while the plugin is also running commands, it will crash cadwork. 
For example, the plugin has cadwork save at the same time the plugin saves information to its own database, and if you try and change any elements while cadwork is saving, it seems to crash every time.
There are a couple other examples that seem to have similar root causes, where people try to do something in cadwork at the same time a plugin is calling cadwork endpoints.

We've been able to fix or get around a couple of the issues so far when we do get the console crash message, but I'd like to get a proper log going so that I can investigate the issue further.

Also, for added context, we are using cwapi3dcpp instead of cwapi3dpython, there just wasn't a discussion board there :P

Thanks for your help!

**Reply by swatkinson** [reply]
I've replicated something similar to this in my plugin, it logs exactly what I need, thanks for your assistance @Brunner246 ! 😀

**Reply by Brunner246** [reply]
Great, I'm glad to hear that 😃

**Comment by Brunner246** [comment]
you could may try to use `std::freopen` 
You could also log the users actions for example with https://github.com/gabime/spdlog to a file. 

are you doing anything in parallel in your plugins ?

**Reply by swatkinson** [reply]
Nope, we tried running the endpoint calls in parallel, but it caused a bunch more issues, so we run them one at a time now.

I'll look into writing my plugins' standard error to a log file, that seems to be the easiest way to get that information logged if there's no console logging built into cadwork.

Thanks for the help and the quick responses! :)

**Comment by roehligas** [comment]
We have this in every of our scripts:

```
"""
This is the Wrapper-script for the Function.

It stores the errorlog as a text file if the execution is crashing.
"""

# Import packages
import time
import os
import sys
import traceback

import utility_controller as uc

##################################################################################################
#                                                                                                #
#                                                                                                #
#                             1.  FUNCTION-SECTION                                               #
#                                                                                                #
#                                                                                                #
##################################################################################################

def time2string(time_in):
    """
    This function converts the time object into a readable string.
    
    
    """
    y= str(curr_time.tm_year).zfill(4)
    M = str(curr_time.tm_mon).zfill(2)
    d = str(curr_time.tm_mday).zfill(2)
    h = str(curr_time.tm_mday).zfill(2)
    m = str(curr_time.tm_min).zfill(2)
    s = str(curr_time.tm_sec).zfill(2)
    
    date_str = y+' '+M+' '+d+'  '+h+' '+m+' '+s+' '
    
    return date_str







##################################################################################################
#                                                                                                #
#                                                                                                #
#                             2.  SCRIPT-SECTION                                                 #
#                                                                                                #
#                                                                                                #
##################################################################################################



try:

    # Add the folder where this script file i stored to system paths
    curr_script_folder = os.path.dirname(__file__)
    sys.path.append(curr_script_folder)
    
    
    ######################################################
    #                                                    #
    #        WRITE ACTUAL TO-DO SEQUENCE HERE            #
    #                                                    #
    ######################################################
    
    
    # Import the actual script
    import GenerateDoorBeams_lib as ess
    
    ess.main()
    
    
    
    ######################################################
    #                                                    #
    #        END ACTUAL TO-DO SECTION                    #
    #                                                    #
    ######################################################




# Catching runtime Exceptions
except Exception as E:
    uc.enable_auto_display_refresh()
    
    # Get the folder where this script is stored
    curr_script_folder = os.path.dirname(__file__)
    
    # Get Error Traceback
    traceback_error = traceback.format_exc()
    
    # Get current Time when error occured
    curr_time = time.localtime()
    timestamp = time2string(curr_time)
    
    
    # Save errormessage in the folder where the script is stored
    filename_errorlog = os.path.join(curr_script_folder,'Errorlog '+timestamp+'.txt')
    log_file = open(filename_errorlog,'w')
    log_file.write(str(traceback_error))
    log_file.close()
    
    # Open the generated file to give the user a feedback
    os.startfile(filename_errorlog)


uc.enable_auto_display_refresh()
```

Which stores automatically errorlog files in the folder of the script.

Makes my life so much easier to keep track of all employees and their bugs

![image](https://github.com/user-attachments/assets/695b341f-4c7d-4f92-99f6-9d61c57bd741)
